### PR TITLE
Show instructions on initial Start up and refactor tabs items

### PR DESCRIPTION
**Pull request by clementparizot**
Work in Progress
Added a page at the start of the application. Still needs to fix the layout.
There's a checkbox to disable this page at the start up, is there a property file for that ?
Also feedbacks are welcomed on the text content.

### CODE REVIEW

**Comment by johnzeringue** [comment]
There's no property for that. Persistent settings is not something we've handled so far. Maybe this isn't the right use case to vet them.

An alternative to the checkbox would be to show the startup page whenever there aren't any datasets.

**Comment by johnzeringue** [comment]
What happens when `TSVTable#getCurrentTable` is called when the instructions tab is open?

Consider the following:

1. A new user opens Topsoil for the first time and is greeted by the instructions tab.
2. Following the guide, he successfully imports data into a new tab.
3. Switching tabs, the user references the instructions again.
4. He tries creating a new chart by clicking `Charts > error ellipse chart`.

Nothing would happen, correct? Of course, the user should have created the chart from the other tab, but I think this example illustrates an architectural problem with having heterogeneous tabs.

I see two solutions:

1. Have the instructions in a separate pane/window.
2. Make the tabs homogeneous again by introducing a super type or a wrapper type for both `TSVTable` and `InstructionsPanel`.

@clementparizot, thoughts?

File: src/main/java/org/cirdles/topsoil/app/TSVTable.java [hunk 1]
```diff
@@ -18,6 +18,7 @@
 import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.Optional;
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TSVTable.java]: Why?

File: src/main/java/org/cirdles/topsoil/app/InstructionsPanel.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+import com.johnzeringue.extendsfx.annotation.ResourceBundle;
+import com.johnzeringue.extendsfx.layout.CustomVBox;
+
+/**
+ *
+ * @author parizotclement
+ */
+@ResourceBundle("Resources")
+public class InstructionsPanel extends CustomVBox<InstructionsPanel> {
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/InstructionsPanel.java]: This can just `extends CustomVBox`. Make sure you understand what the generic type is for.

File: src/main/java/org/cirdles/topsoil/app/TabContents.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            self.instructions = Optional.of(instructions);
+        });
+    }
+
+    @FXML
+    private void initialize() {
+        getTable().ifPresent(getChildren()::setAll);
+        getInstructions().ifPresent(getChildren()::setAll);
+
+        setVgrow(table.get(), Priority.ALWAYS);
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TabContents.java]: This would be better in the FXML.

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 4]
```diff
@@ -99,8 +104,15 @@ private void initialize() {
                 String.format("%s [%s]", applicationName, applicationVersion));
     }
 
-    Optional<TSVTable> getCurrentTable() {
-        return getCurrentTab().map(tab -> (TSVTable) tab.getContent());
+    Optional<TabContents> getCurrentTable() {
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Should still return `Optional<TSVTable>` unless you feel this method's no longer necessary.

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 5]
```diff
@@ -99,8 +104,15 @@ private void initialize() {
...
+        if (!getCurrentTab().isPresent()) {
+            //No tab
+            warningPrompt("There's no table open.");
+            return Optional.empty();
+
+        } else {
+            return getCurrentTab().map(tab -> (TabContents) tab.getContent());
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Maybe introduce `getCurrentTabContents()`?

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 6]
```diff
@@ -132,20 +161,29 @@ Tab createTab() {
 
     @FXML
     void saveDataTable() {
-        TextInputDialog textInputDialog = new TextInputDialog();
+        if (getCurrentTable().isPresent()) {
+            if (getCurrentTable().get().getTable().isPresent()) {
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Please refactor.

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 7]
```diff
@@ -132,20 +161,29 @@ Tab createTab() {
...
+                            .resolve(datasetName + ".tsv");
 
-            getCurrentTable().ifPresent(table -> table.saveToPath(datasetPath));
-            getCurrentTab().ifPresent(tab -> tab.setText(datasetName));
-        });
+                    getCurrentTab().ifPresent(tab -> tab.setText(datasetName));
 
-        // reload
-        reloadDatasetMenu();
+                    getCurrentTable().get().getTable().get().saveToPath(datasetPath);
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Refactor.

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 8]
```diff
@@ -299,29 +343,66 @@ void initializeAndShow(Chart chart, Dataset dataset) {
     }
 
     private void initializeAndShow(JavaScriptChart javaScriptChart) {
-        getCurrentTable().map(TSVTable::getDataset).ifPresent(dataset -> {
+        getCurrentTable().get().getTable().map(TSVTable::getDataset).ifPresent(dataset -> {
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Refactor.

File: src/main/java/org/cirdles/topsoil/app/TabContents.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+            self.instructions = Optional.of(instructions);
+        });
+    }
+
+    @FXML
+    private void initialize() {
+        getTable().ifPresent(getChildren()::setAll);
+        getInstructions().ifPresent(getChildren()::setAll);
+
+        //setVgrow(table.get(), Priority.ALWAYS);
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TabContents.java]: Please remove this entirely. I don't like commented out code.

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 10]
```diff
@@ -299,29 +343,66 @@ void initializeAndShow(Chart chart, Dataset dataset) {
...
         });
     }
 
     @FXML
     void createScatterplot() {
-        initializeAndShow(new ScatterplotChart());
+        if (getCurrentTable().isPresent()) {
+            getCurrentTable().ifPresent(table -> {
+                initializeAndShow(new ScatterplotChart());
+            });
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: It's okay to call `Optional#get` after `Optional#isPresent`.

File: src/main/resources/org/cirdles/topsoil/app/TabContents.fxml [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+<?import java.util.*?>
+<?import javafx.scene.*?>
+<?import javafx.scene.control.*?>
+<?import javafx.scene.layout.*?>
+
+<fx:root xmlns:fx="http://javafx.com/fxml/1"
+         type="org.cirdles.topsoil.app.TabContents"
+         alignment="CENTER">
+    
+    <StackPane fx:id="table" AnchorPane.topAnchor="0.0" AnchorPane.bottomAnchor="0.0" AnchorPane.leftAnchor="0.0" AnchorPane.rightAnchor="0.0" ></StackPane>
```
Comment (johnzeringue) [inline src/main/resources/org/cirdles/topsoil/app/TabContents.fxml]: If you want a `StackPane`, why isn't `TabContents` one?

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 12]
```diff
@@ -104,10 +105,25 @@ private void initialize() {
...
+            return Optional.empty();
+
+        } else {
+            return getCurrentTab().map(tab -> (TabContents) tab.getContent());
+        }
     }
 
     Optional<TSVTable> getCurrentTable() {
-        return getCurrentTab().map(tab -> (TSVTable) tab.getContent());
+        return getCurrentTabContents().map(table -> table.getTable().orElse(null));
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: This is not good. Please look at how we've used `Optional` elsewhere or maybe check for resources online. Hint: Use `flatMap`.

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java (hunk 5 above)
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Any thoughts?

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 13]
```diff
@@ -115,6 +131,22 @@ void createDataTable() {
...
+
+        InstructionsPanel instructions = new InstructionsPanel();
+        TabContents content = new TabContents(instructions);
+        instructionsTab.setContent(content);
+
+        dataTableTabPane.getTabs().add(instructionsTab);
+
+        // focus on new tab
+        SelectionModel<Tab> selectionModel = dataTableTabPane.getSelectionModel();
+        selectionModel.select(instructionsTab);
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: This logic should be shared between `createInstructionsTab` and `createTable`.

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 14]
```diff
@@ -310,29 +354,60 @@ void initializeAndShow(Chart chart, Dataset dataset) {
...
+        }
     }
 
     @FXML
     void pasteFromClipboard() {
-        getCurrentTable().ifPresent(TSVTable::pasteFromClipboard);
+        getCurrentTable()
+                .ifPresent(
+                        currentTable -> currentTable.pasteFromClipboard()
+                );
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Why did you change this? Do you not understand the other syntax?

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 15]
```diff
@@ -235,7 +278,8 @@ public void changed(ObservableValue<? extends Scene> observableScene, Scene oldS
     }
 
     public TSVTable createTable() {
-        return (TSVTable) createTab().getContent();
+        TabContents content = (TabContents) createTab().getContent();
+        return content.getTable().get();
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: This has to be rewritten for `TabContents`. `createInstructionsTab` and this should share the logic in `createTab`. Here, `createTab` is creating the table.

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 16]
```diff
@@ -310,29 +354,60 @@ void initializeAndShow(Chart chart, Dataset dataset) {
...
     }
 
     @FXML
     void emptyTable() {
-        getCurrentTable().ifPresent(TSVTable::clear);
+        if (getCurrentTable().isPresent()) {
+            getCurrentTable().get().clear();
+        } else {
+            //Not a TSVTable
+            warningPrompt("You can't clear this table.");
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: If `getCurrentTable` isn't present, why are you referring to it?

File: src/main/resources/org/cirdles/topsoil/app/InstructionsPanel.fxml [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+    <Text textAlignment="CENTER" style="-fx-font-size: 20pt;">
+        Welcome to Topsoil
+    </Text>
+    
+    <VBox style="-fx-font-size: 12pt;" spacing="20" >
+        <padding>
+            <Insets left="50" top="50" right="50" bottom="50" />
+        </padding>
+        <Label wrapText="true" 
+               text="To start, go to 'File' > 'Import from TSV', and select the .tsv file you want to load."/>
```
Comment (johnzeringue) [inline src/main/resources/org/cirdles/topsoil/app/InstructionsPanel.fxml]: Long text like this is better in a `.properties` file. Make `InstructionsPanel.properties` in this same package.

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java (hunk 16 above)
Comment (clementparizot) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: getCurrentTable() returns a TSVTable. If there's no TSVTable it means that it's the instructions panel.

Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: And if it's the instructions panel, it warns "can't clear table", hence the confusion.

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 18]
```diff
@@ -104,15 +105,30 @@ private void initialize() {
...
+            return Optional.empty();
+
+        } else {
+            return getCurrentTab().map(tab -> (TabContents) tab.getContent());
+        }
     }
 
     Optional<TSVTable> getCurrentTable() {
-        return getCurrentTab().map(tab -> (TSVTable) tab.getContent());
+        return getCurrentTabContents().flatMap(table -> table.getTable());
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Use `::`.

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 19]
```diff
@@ -137,22 +149,47 @@ Tab createTab() {
...
 
+    TSVTable createTable() {
+        Tab tab = createTab();
+
+        TSVTable dataTable = new TSVTable();
+        dataTable.setPlaceholder(new EmptyTablePlaceholder(dataTable));
+        TabContents content = new TabContents(dataTable);
+        tab.setContent(content);
+
+        return content.getTable().get();
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: `return dataTable;`?

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 20]
```diff
@@ -137,22 +149,47 @@ Tab createTab() {
...
+    }
+
+    void createInstructionsPanel() {
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Should this signature match `createTable` more?

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 21]
```diff
@@ -310,19 +344,29 @@ void initializeAndShow(Chart chart, Dataset dataset) {
...
+        }
     }
 
     @FXML
     void createErrorEllipseChart() {
-        initializeAndShow(new ErrorEllipseChart());
+        if (getCurrentTable().isPresent()) {
+                initializeAndShow(new ErrorEllipseChart());
+        } else {
+            //Not a TSVTable
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Then why do we call it one in the warning.

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java [hunk 22]
```diff
@@ -310,19 +344,29 @@ void initializeAndShow(Chart chart, Dataset dataset) {
...
+            warningPrompt("You can't draw a chart from this table.\nPlease select a valid table first.");
+        }
```
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Can we remove the duplication in the two methods above?

File: src/main/resources/Resources.properties [hunk 23]
```diff
@@ -70,3 +70,8 @@ rhoLabelText = \u03c1
 wetherillTypeLabel      = Wetherill
 teraWasserburgTypeLabel = Tera-Waserburg
 noneTypeLabel           = None
+
+# Instructions Panel
+importContent = To start, go to 'File' > 'Import from TSV', and select the .tsv file you want to load.
+createTab = Alternatively, you can click on 'Create a Data Table', copy the data directly from your .tsv file and then click on 'Paste from Clipboard'.
+drawChart = Once you've imported your data, you can go to 'Charts' menu and select a chart.
```
Comment (johnzeringue) [inline src/main/resources/Resources.properties]: "Long text like this is better in a `.properties` file. Make `InstructionsPanel.properties` in this same package."

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java (hunk 20 above)
Comment (clementparizot) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: What do you mean ?

Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: One has a return type and one doesn't. Which is better?

File: src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java (hunk 5 above)
Comment (johnzeringue) [inline src/main/java/org/cirdles/topsoil/app/TopsoilMainWindow.java]: Address the fact that this method makes no sense and I'll merge.

**Comment by johnzeringue** [comment]
This pull request is being closed as @clementparizot is no longer active in the project and this code is out of date. @clementparizot, thanks again for your efforts!
